Provide an Option to Use Path-Style-Access with S3 Repo

original-brownbear opened the pull request:
* As discussed, added the option to use path style access back again and
deprecated it.
* Defaulted to falling back to the SDK defaults
* Added warning to docs + breaking changes entry

* Closes #41816

original-brownbear committed:
Provide an Option to Use Path-Style-Access with S3 Repo

* As discussed, added the option to use path style access back again and
deprecated it.
* Defaulted to `false`
* Added warning to docs

* Closes #41816

original-brownbear committed:
add breaking changes entry

elasticmachine commented:
Pinging @elastic/es-distributed

original-brownbear commented on plugins/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3ClientSettings.java:
```diff
@@ -95,6 +95,10 @@
 
+    /** Whether the s3 client should use path style access. */
+    static final Setting.AffixSetting<Boolean> USE_PATH_STYLE_ACCESS = Setting.affixKeySetting(PREFIX, "path_style_access",
+        key -> Setting.boolSetting(key, false, Property.NodeScope, Property.Deprecated));
```
Now that AWS changed course on this one and has announced continued support for this access pattern for existing/older buckets (https://aws.amazon.com/blogs/aws/amazon-s3-path-deprecation-plan-the-rest-of-the-story), maybe there's no need to change the default here before `8.0`?

ywelsch commented on plugins/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3ClientSettings.java:
```diff
@@ -127,9 +131,13 @@
     final boolean throttleRetries;
 
+    /** Whether the s3 client should use path style access. */
+    final boolean pathStyleAccess;
```
The AWS SDK uses this as a ternary setting (true, false, null), with null being the default, which enables the path-style setting based on the configured endpoint. I think we should do treat this setting exactly the same way, allowing path-style setting both being explicitly enabled and disabled.

original-brownbear replied:
Sounds good, so we break behavior after all by changing the default to `null`? See my comment above, this may not be necessary (at least in 7.x)?

DaveCTurner replied:
It's unclear what the behaviour of `setPathStyleAccessEnabled(Boolean.FALSE)` is from the SDK docs. I think we're doing the right thing here, either enabling it or leaving it unset.

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into fix-s3-path-styl-access

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into fix-s3-path-styl-access

original-brownbear committed:
default to SDK defaults

original-brownbear commented:
@ywelsch Adjusted this to a ternary setting now and default to the SDK behavior as discussed. Take a look when you have a chance :)

ywelsch commented on docs/reference/migration/migrate_8_0/snapshots.asciidoc:
```diff
@@ -28,3 +28,15 @@ This change will affect both newly created repositories and existing repositorie
+[float]
+==== The S3 repository plugin uses the DNS style access pattern by default
+
+Starting in version 7.3 using the path style access pattern with the S3 repository is deprecated.
```
I'm not sure we should deprecate this at this time. AWS's SDK is going to support path-style access for a very long time to come (even after Sept. 2020, given that the SDK will still be used to connect to older repos).

original-brownbear replied:
Same point as in the above, why make a breaking change then (that we know for a fact will break some environments)?

ywelsch replied:
The focus of this section should be on the fact that the behavior changes between 7.2-  and 7.3+, going back to the 6.x behavior, and that the 7.0-7.2 behavior can be restored by setting this new setting. The deprecation of that new setting is secondary. As said, given that we don't have a time frame for removing this setting, I would prefer to defer the deprecation to a later time.

ywelsch commented on plugins/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3ClientSettings.java:
```diff
@@ -95,6 +96,12 @@
+    static final Setting.AffixSetting<S3ClientSettings.PathStyleAccess> USE_PATH_STYLE_ACCESS = Setting.affixKeySetting(
+        PREFIX, "path_style_access",
+        key -> new Setting<>(key, s -> "default", s -> Boolean.parseBoolean(s) ? PathStyleAccess.ENABLED : PathStyleAccess.DISABLED,
+            Property.NodeScope, Property.Deprecated));
```
As said above, I don't think we should log deprecation warnings.

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into fix-s3-path-styl-access

original-brownbear commented on plugins/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3ClientSettings.java:
```diff
@@ -95,6 +96,12 @@
+    /** Whether the s3 client should use path style access. */
+    static final Setting.AffixSetting<S3ClientSettings.PathStyleAccess> USE_PATH_STYLE_ACCESS = Setting.affixKeySetting(
+        PREFIX, "path_style_access",
+        key -> new Setting<>(key, s -> "default", s -> Boolean.parseBoolean(s) ? PathStyleAccess.ENABLED : PathStyleAccess.DISABLED,
```
My idea was to simply only allow parsing `true` and `false` here and let the `DEFAULT` represent the setting not being set (I didn't see much value in allowing to explicitly set "default" or "null"). Not a good idea? :)

original-brownbear commented on docs/plugins/repository-s3.asciidoc:
```diff
@@ -145,6 +145,18 @@ settings belong in the `elasticsearch.yml` file.
+   AWS S3 will stop supporting the path style access pattern for new buckets from
+   https://aws.amazon.com/blogs/aws/amazon-s3-path-deprecation-plan-the-rest-of-the-story[September 30th, 2020].
+   Any setups using the path style access pattern should be switched to the domain style access pattern as soon as possible to ensure
+   continued compatibility with the S3 repository plugin.
```
Isn't toning it down a little inconsistent with us now making a breaking change in `7.x`?
My thinking was that we're doing this breaking change because we fear that future versions of the SDK won't support path style access. If we don't think this is an imminent enough issue to strongly push users towards adjusting their environment asap then why make a breaking change and not just wait for `8.0`?

ywelsch replied:
> Isn't toning it down a little inconsistent with us now making a breaking change in 7.x?

It's just going back to the pre-7.0 behavior, which was the more sensible default, especially given the current strategy of Amazon not to support path-style access in the future anymore, and there being no way to explicitly force-disable it in the SDK (i.e. it would still use path-style access with IP-based endpoint). This here makes it sound like we would break Minio support with path-style access at any moment in time, which sounds perhaps a bit overdramatic.

original-brownbear committed:
CR: Adjust docs

original-brownbear commented:
@ywelsch all points addressed :), left some questions though.

original-brownbear commented:
Jenkins run elasticsearch-ci/1

ywelsch commented on docs/reference/migration/migrate_8_0/snapshots.asciidoc:
```diff
@@ -28,3 +28,15 @@ This change will affect both newly created repositories and existing repositorie
+
+Starting in version 7.3 using the path style access pattern with the S3 repository is deprecated.
+In versions 7.0, 7.1, and 7.2 the S3 repository plugin was exclusively using the path style access pattern. This is a breaking
+change for deployments that do not also allow for the DNS style access pattern. For short-term compatibility with these deployments users
```
only those that are not on AWS and that do not use IP-address based endpoint.

ywelsch reviewed (changes requested):
I have added a few more replies. Also, still missing tests.

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into fix-s3-path-styl-access

original-brownbear commented on docs/plugins/repository-s3.asciidoc:
```diff
@@ -145,6 +145,19 @@ settings belong in the `elasticsearch.yml` file.
+`path_style_access`::
+
+   Whether to use the path style access pattern. If `true`, the path style access pattern will be used. If set to`false`, DNS style access will
+   be used. Defaults to letting the AWS Java SDK decide whether to use the path style access pattern dynamically
```
RIght ... I should've looked on step deeper `S3ClientOptions` uses `boolean` already ... will simplify this PR then :)

original-brownbear committed:
boolean setting

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into fix-s3-path-styl-access

original-brownbear committed:
CR: tone down language :)

original-brownbear committed:
tone it down some more

original-brownbear commented:
@ywelsch thanks for taking another look :) Sorry for the back and forth here, I now way toned down the language, moved to a boolean setting and added a test :) Should be good for another review I hope.

ywelsch commented on docs/plugins/repository-s3.asciidoc:
```diff
@@ -145,6 +145,17 @@ settings belong in the `elasticsearch.yml` file.
+   (See https://docs.aws.amazon.com/AWSJavaSDK/latest/javadoc/com/amazonaws/services/s3/AmazonS3Builder.html#setPathStyleAccessEnabled-java.lang.Boolean-[AWS documentation] for details).
+   Defaults to `false`.
+
+NOTE: In versions `7.0` and `7.1`, all bucket operations were using the path style access pattern in every situation without an option to
```
7.0, 7.1 and 7.2

ywelsch commented on docs/plugins/repository-s3.asciidoc:
```diff
@@ -145,6 +145,17 @@ settings belong in the `elasticsearch.yml` file.
 
+`path_style_access`::
+
+   Whether to use the path style access pattern. If `true`, the path style access pattern will be used. If set to`false`,
```
If set to `false`, (space missing)

also perhaps write this as

If left to `false`,

to make it clearer that setting this to false just uses the default behavior.

ywelsch replied:
s/Whether to use the path style access pattern/Whether to force the use of the path style access pattern/

ywelsch commented on docs/plugins/repository-s3.asciidoc:
```diff
@@ -145,6 +145,17 @@ settings belong in the `elasticsearch.yml` file.
+`path_style_access`::
+
+   Whether to use the path style access pattern. If `true`, the path style access pattern will be used. If set to`false`,
+   the AWS Java SDK decide whether to use the path style access pattern dynamically
```
path style access will be automatically determined by the AWS Java SDK.

ywelsch reviewed (approved):
I've left a few smaller doc changes, but looks good o.w.
@DaveCTurner can you have a look as well?

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into fix-s3-path-styl-access

original-brownbear committed:
CR: fix docs

original-brownbear commented:
@DaveCTurner ping :)

DaveCTurner committed:
Docs fixes post 7.3 FF

DaveCTurner committed:
Merge branch 'master' into fix-s3-path-styl-access

DaveCTurner committed:
Wording tweak

DaveCTurner reviewed (approved):
Sorry, somehow I missed this past the 7.3 feature freeze. In repentance, I've updated the docs to refer back to 7.0-7.3 and cleaned up the wording and formatting a bit too, see f76a14e. This LGTM now.

original-brownbear commented:
thanks @DaveCTurner !

sonujatav35 commented:
Hi
Anywork around for Elasticsearch v6.2. My s3 compatible storage only accept path_style_access request and this option is not exist in ES v6. How can i make work so s3 plugin can use path_style_access instead of using virtual style access.

DaveCTurner commented:
@sonujatav35 please ask questions like this on the [Elasticsearch forum](https://discuss.elastic.co/c/elasticsearch). This allows us to use GitHub for verified bug reports, feature requests, and pull requests. 6.2 is almost 2 years past [EOL](https://www.elastic.co/support/eol), you should upgrade as a matter of urgency.
